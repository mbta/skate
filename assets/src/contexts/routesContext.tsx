import React, { createContext, ReactElement, useContext } from "react"
import { Route, RouteId } from "../schedule"

const RoutesContext = createContext<Route[] | null>(null)

export const useRoute = (routeId: RouteId | null | undefined): Route | null => {
  const routes: Route[] | null = useContext(RoutesContext)
  return (routes || []).find((route) => route.id === routeId) || null
}

export const RoutesProvider = ({
  routes,
  children,
}: {
  routes: Route[] | null
  children: ReactElement<HTMLElement>
}) => {
  return (
    <RoutesContext.Provider value={routes}>{children}</RoutesContext.Provider>
  )
}

export default RoutesContext
