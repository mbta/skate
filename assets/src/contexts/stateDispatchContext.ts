import { createContext } from "react"
import { Dispatch, initialState, State } from "../state"

type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export default StateDispatchContext
