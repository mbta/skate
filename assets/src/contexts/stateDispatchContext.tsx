import React, { createContext, ReactElement } from "react"
import { Dispatch, initialState, State } from "../state"

export type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
export const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export const StateDispatchProvider = ({
  state,
  dispatch,
  children,
}: {
  state: State
  dispatch: Dispatch
  children: ReactElement<HTMLElement>
}) => {
  return (
    <StateDispatchContext.Provider value={[state, dispatch]}>
      {children}
    </StateDispatchContext.Provider>
  )
}
