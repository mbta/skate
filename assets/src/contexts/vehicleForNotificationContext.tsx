import React, { createContext, ReactElement } from "react"
import { VehicleOrGhost } from "../realtime"

const VehicleForNotificationContext = createContext<
  VehicleOrGhost | null | undefined
>(null)

export const VehicleForNotificationProvider = ({
  vehicleForNotification,
  children,
}: {
  vehicleForNotification: VehicleOrGhost | null | undefined
  children: ReactElement<HTMLElement>
}) => {
  return (
    <VehicleForNotificationContext.Provider value={vehicleForNotification}>
      {children}
    </VehicleForNotificationContext.Provider>
  )
}

export default VehicleForNotificationContext
