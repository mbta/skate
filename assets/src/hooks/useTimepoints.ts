import { useEffect, useState } from "react"
import { fetchTimepointsForRoute } from "../api"
import { RouteId, Timepoint, TimepointsByRouteId } from "../skate"

const useTimepoints = (selectedRouteIds: RouteId[]): TimepointsByRouteId => {
  const [timepointsByRouteId, setTimepointsByRouteId] = useState<
    TimepointsByRouteId
  >({})

  const setLoadingTimepointsForRoute = (routeId: RouteId): void => {
    setTimepointsByRouteId({
      ...timepointsByRouteId,
      [routeId]: null,
    })
  }

  const setTimepointsForRoute = (
    routeId: RouteId,
    timepoints: Timepoint[]
  ): void => {
    setTimepointsByRouteId({
      ...timepointsByRouteId,
      [routeId]: timepoints,
    })
  }

  useEffect(() => {
    selectedRouteIds.forEach((routeId: RouteId) => {
      if (!(routeId in timepointsByRouteId)) {
        setLoadingTimepointsForRoute(routeId)

        fetchTimepointsForRoute(routeId).then((newTimepoints: Timepoint[]) =>
          setTimepointsForRoute(routeId, newTimepoints)
        )
      }
    })
  }, [selectedRouteIds, timepointsByRouteId])

  return timepointsByRouteId
}

export default useTimepoints
