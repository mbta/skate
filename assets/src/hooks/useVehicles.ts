import { Channel, Socket } from "phoenix"
import { Dispatch as ReactDispatch, useEffect, useReducer } from "react"
import { HeadwaySpacing } from "../models/vehicleStatus"
import {
  DataDiscrepancy,
  Vehicle,
  VehicleScheduledLocation,
  VehiclesForRoute,
  VehicleStatus,
  VehicleStopStatus,
  VehicleTimepointStatus,
} from "../realtime.d"
import { ByRouteId, DirectionId, RouteId } from "../schedule.d"

interface DataDiscrepancyData {
  attribute: string
  sources: DataDiscrepancySourceData[]
}

interface DataDiscrepancySourceData {
  id: string
  value: string
}

type RawHeadwaySpacing =
  | "very_bunched"
  | "bunched"
  | "ok"
  | "gapped"
  | "very_gapped"
  | null

interface VehicleScheduledLocationData {
  direction_id: DirectionId
  timepoint_status: VehicleTimepointStatusData
}

interface VehicleStopStatusData {
  status: VehicleStatus
  stop_id: string
  stop_name: string
}

interface VehicleTimepointStatusData {
  timepoint_id: string
  fraction_until_timepoint: number
}

interface VehicleData {
  id: string
  label: string
  run_id: string
  timestamp: number
  latitude: number
  longitude: number
  direction_id: DirectionId
  route_id: RouteId
  trip_id: string
  headsign: string | null
  via_variant: string | null
  operator_id: string
  operator_name: string
  bearing: number
  speed: number
  block_id: string
  headway_secs: number
  headway_spacing: RawHeadwaySpacing
  previous_vehicle_id: string
  schedule_adherence_secs: number
  schedule_adherence_string: string
  scheduled_headway_secs: number
  is_off_course: boolean
  is_laying_over: boolean
  layover_departure_time: number | null
  block_is_active: boolean
  sources: string[]
  data_discrepancies: DataDiscrepancyData[]
  stop_status: VehicleStopStatusData
  timepoint_status: VehicleTimepointStatusData | null
  scheduled_location: VehicleScheduledLocationData | null
}

interface VehiclesForRouteData {
  on_route_vehicles: VehicleData[]
  incoming_vehicles: VehicleData[]
}

interface State {
  channelsByRouteId: ByRouteId<Channel>
  vehiclesByRouteId: ByRouteId<VehiclesForRoute>
}

const initialState: State = {
  channelsByRouteId: {},
  vehiclesByRouteId: {},
}

interface SetChannelForRouteAction {
  type: "SET_CHANNEL"
  payload: {
    routeId: RouteId
    channel: Channel
  }
}

const headwaySpacing = (raw: RawHeadwaySpacing): HeadwaySpacing | null => {
  switch (raw) {
    case null:
      return null

    case "very_bunched":
      return HeadwaySpacing.VeryBunched

    case "bunched":
      return HeadwaySpacing.Bunched

    case "ok":
      return HeadwaySpacing.Ok

    case "gapped":
      return HeadwaySpacing.Gapped

    case "very_gapped":
      return HeadwaySpacing.VeryGapped
  }
}

const setChannelForRoute = (
  routeId: RouteId,
  channel: Channel
): SetChannelForRouteAction => ({
  type: "SET_CHANNEL",
  payload: { routeId, channel },
})

interface SetVehiclesForRouteAction {
  type: "SET_VEHICLES_FOR_ROUTE"
  payload: {
    routeId: RouteId
    vehiclesForRoute: VehiclesForRoute
  }
}

const setVehiclesForRoute = (
  routeId: RouteId,
  vehiclesForRoute: VehiclesForRoute
): SetVehiclesForRouteAction => ({
  type: "SET_VEHICLES_FOR_ROUTE",
  payload: {
    routeId,
    vehiclesForRoute,
  },
})

interface RemoveRouteAction {
  type: "REMOVE_ROUTE"
  payload: {
    routeId: RouteId
  }
}

const removeRoute = (routeId: RouteId): RemoveRouteAction => ({
  type: "REMOVE_ROUTE",
  payload: { routeId },
})

type Action =
  | SetChannelForRouteAction
  | SetVehiclesForRouteAction
  | RemoveRouteAction

type Dispatch = ReactDispatch<Action>

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "SET_CHANNEL":
      return {
        ...state,
        channelsByRouteId: {
          ...state.channelsByRouteId,
          [action.payload.routeId]: action.payload.channel,
        },
      }
    case "SET_VEHICLES_FOR_ROUTE":
      return {
        ...state,
        vehiclesByRouteId: {
          ...state.vehiclesByRouteId,
          [action.payload.routeId]: action.payload.vehiclesForRoute,
        },
      }
    case "REMOVE_ROUTE":
      const {
        [action.payload.routeId]: _channel,
        ...channelsWithoutRouteId
      } = state.channelsByRouteId
      const {
        [action.payload.routeId]: _vehiclesForRoute,
        ...vehiclesByRouteIdWithoutRouteId
      } = state.vehiclesByRouteId
      return {
        ...state,
        channelsByRouteId: channelsWithoutRouteId,
        vehiclesByRouteId: vehiclesByRouteIdWithoutRouteId,
      }
    default:
      return state
  }
}

const dataDiscrepanciesFromData = (
  dataDiscrepancies: DataDiscrepancyData[]
): DataDiscrepancy[] =>
  dataDiscrepancies.map(dataDiscrepancy => ({
    attribute: dataDiscrepancy.attribute,
    sources: dataDiscrepancy.sources,
  }))

const vehicleScheduledLocationFromData = (
  vehicleScheduledLocationData: VehicleScheduledLocationData
): VehicleScheduledLocation => ({
  directionId: vehicleScheduledLocationData.direction_id,
  timepointStatus: vehicleTimepointStatusFromData(
    vehicleScheduledLocationData.timepoint_status
  ),
})

const vehicleStopStatusFromData = (
  vehicleStopStatusData: VehicleStopStatusData
): VehicleStopStatus => ({
  status: vehicleStopStatusData.status,
  stopId: vehicleStopStatusData.stop_id,
  stopName: vehicleStopStatusData.stop_name,
})

const vehicleTimepointStatusFromData = (
  vehicleTimepointStatusData: VehicleTimepointStatusData
): VehicleTimepointStatus => ({
  timepointId: vehicleTimepointStatusData.timepoint_id,
  fractionUntilTimepoint: vehicleTimepointStatusData.fraction_until_timepoint,
})

const vehicleFromData = ({ isOnRoute }: { isOnRoute: boolean }) => (
  vehicleData: VehicleData
): Vehicle => ({
  id: vehicleData.id,
  label: vehicleData.label,
  runId: vehicleData.run_id,
  timestamp: vehicleData.timestamp,
  latitude: vehicleData.latitude,
  longitude: vehicleData.longitude,
  directionId: vehicleData.direction_id,
  routeId: vehicleData.route_id,
  tripId: vehicleData.trip_id,
  headsign: vehicleData.headsign,
  viaVariant: vehicleData.via_variant,
  operatorId: vehicleData.operator_id,
  operatorName: vehicleData.operator_name,
  bearing: vehicleData.bearing,
  speed: vehicleData.speed,
  blockId: vehicleData.block_id,
  headwaySecs: vehicleData.headway_secs,
  headwaySpacing: headwaySpacing(vehicleData.headway_spacing),
  previousVehicleId: vehicleData.previous_vehicle_id,
  scheduleAdherenceSecs: vehicleData.schedule_adherence_secs,
  scheduleAdherenceString: vehicleData.schedule_adherence_string,
  scheduledHeadwaySecs: vehicleData.scheduled_headway_secs,
  isOffCourse: vehicleData.is_off_course,
  isLayingOver: vehicleData.is_laying_over,
  layoverDepartureTime: vehicleData.layover_departure_time,
  blockIsActive: vehicleData.block_is_active,
  dataDiscrepancies: dataDiscrepanciesFromData(vehicleData.data_discrepancies),
  stopStatus: vehicleStopStatusFromData(vehicleData.stop_status),
  timepointStatus:
    vehicleData.timepoint_status &&
    vehicleTimepointStatusFromData(vehicleData.timepoint_status),
  scheduledLocation:
    vehicleData.scheduled_location &&
    vehicleScheduledLocationFromData(vehicleData.scheduled_location),
  isOnRoute,
})

const vehiclesForRouteFromData = (
  vehiclesForRouteData: VehiclesForRouteData
): VehiclesForRoute => ({
  onRouteVehicles: vehiclesForRouteData.on_route_vehicles.map(
    vehicleFromData({ isOnRoute: true })
  ),
  incomingVehicles: vehiclesForRouteData.incoming_vehicles.map(
    vehicleFromData({ isOnRoute: false })
  ),
})

const subscribe = (
  socket: Socket,
  routeId: RouteId,
  dispatch: Dispatch
): Channel => {
  const handleVehicles = (vehiclesForRouteData: VehiclesForRouteData) => {
    const vehiclesForRoute = vehiclesForRouteFromData(vehiclesForRouteData)
    dispatch(setVehiclesForRoute(routeId, vehiclesForRoute))
  }

  const topic = `vehicles:${routeId}`
  const channel = socket.channel(topic)

  channel.on("vehicles", handleVehicles)

  // Reload our session if the auth has expired
  channel.on("auth_expired", () => {
    window.location.reload(true)
  })

  channel
    .join()
    .receive("ok", handleVehicles)
    // tslint:disable-next-line: no-console
    .receive("error", ({ reason }) => console.error("join failed", reason))
    // tslint:disable-next-line: no-console
    .receive("timeout", () => console.error("join timeout"))
  return channel
}

const useVehicles = (
  socket: Socket | undefined,
  selectedRouteIds: RouteId[]
): ByRouteId<VehiclesForRoute> => {
  const [state, dispatch] = useReducer(reducer, initialState)
  const { channelsByRouteId, vehiclesByRouteId } = state

  useEffect(() => {
    if (socket) {
      // Unsubscribe from any routes we don't care about anymore
      Object.entries(channelsByRouteId).forEach(([routeId, channel]) => {
        if (!selectedRouteIds.includes(routeId)) {
          channel.leave()
          dispatch(removeRoute(routeId))
        }
      })

      // Subscribe to any routes we're not already subscribed to
      selectedRouteIds.forEach((routeId: RouteId) => {
        if (!(routeId in channelsByRouteId)) {
          const channel = subscribe(socket, routeId, dispatch)
          dispatch(setChannelForRoute(routeId, channel))
        }
      })
    }
  }, [socket, selectedRouteIds])

  return vehiclesByRouteId
}

export default useVehicles
