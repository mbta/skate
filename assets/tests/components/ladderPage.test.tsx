import { mount } from "enzyme"
import React from "react"
import renderer from "react-test-renderer"
import LadderPage, {
  findRouteById,
  findSelectedVehicleOrGhost,
} from "../../src/components/ladderPage"
import { RoutesProvider } from "../../src/contexts/routesContext"
import { StateDispatchProvider } from "../../src/contexts/stateDispatchContext"
import { VehicleForNotificationProvider } from "../../src/contexts/vehicleForNotificationContext"
import useTimepoints from "../../src/hooks/useTimepoints"
import useVehicles from "../../src/hooks/useVehicles"
import { HeadwaySpacing } from "../../src/models/vehicleStatus"
import {
  Ghost,
  Notification,
  Vehicle,
  VehicleOrGhost,
} from "../../src/realtime"
import { ByRouteId, Route, TimepointsByRouteId } from "../../src/schedule.d"
import { initialState, State } from "../../src/state"

jest.mock("../../src/hooks/useTimepoints", () => ({
  __esModule: true,
  default: jest.fn(() => ({})),
}))
jest.mock("../../src/hooks/useVehicles", () => ({
  __esModule: true,
  default: jest.fn(() => ({})),
}))
jest.mock("../../src/hooks/useVehicleForNotification", () => ({
  __esModule: true,
  default: jest.fn(() => undefined),
}))

const mockDispatch = jest.fn()

describe("LadderPage", () => {
  test("renders the empty state", () => {
    const tree = renderer.create(<LadderPage />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("renders with routes", () => {
    const mockState = { ...initialState, selectedRouteIds: ["1"] }
    const tree = renderer
      .create(
        <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
          <RoutesProvider routes={routes}>
            <LadderPage />
          </RoutesProvider>
        </StateDispatchProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("renders with selectedRoutes in different order than routes data", () => {
    const mockState = { ...initialState, selectedRouteIds: ["28", "1"] }
    const tree = renderer
      .create(
        <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
          <RoutesProvider routes={routes}>
            <LadderPage />
          </RoutesProvider>
        </StateDispatchProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("renders with timepoints", () => {
    const mockState = { ...initialState, selectedRouteIds: ["28", "1"] }
    ;(useTimepoints as jest.Mock).mockImplementationOnce(
      () => timepointsByRouteId
    )
    const tree = renderer
      .create(
        <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
          <RoutesProvider routes={routes}>
            <LadderPage />
          </RoutesProvider>
        </StateDispatchProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("renders with vehicles and selected vehicles", () => {
    const vehicle: VehicleOrGhost = {
      id: "ghost-id",
      directionId: 0,
      routeId: "1",
      tripId: "trip",
      headsign: "headsign",
      blockId: "block",
      runId: null,
      viaVariant: null,
      layoverDepartureTime: null,
      scheduledTimepointStatus: {
        timepointId: "hhgat",
        fractionUntilTimepoint: 0.0,
      },
      routeStatus: "on_route",
      blockWaivers: [],
    }
    ;(useVehicles as jest.Mock).mockImplementationOnce(() => ({
      ["1"]: [vehicle],
    }))
    const mockState = {
      ...initialState,
      selectedRouteIds: ["1"],
      selectedVehicleId: "ghost-id",
    }
    const tree = renderer
      .create(
        <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
          <RoutesProvider routes={routes}>
            <LadderPage />
          </RoutesProvider>
        </StateDispatchProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("shows VPP from a selected notification", () => {
    const notification: Notification = { runIds: ["run_id"] } as Notification
    const state: State = { ...initialState, selectedNotification: notification }
    const wrapper = mount(
      <StateDispatchProvider state={state} dispatch={jest.fn()}>
        <VehicleForNotificationProvider
          vehicleForNotification={notificationVehicle}
        >
          <LadderPage />
        </VehicleForNotificationProvider>
      </StateDispatchProvider>
    )
    expect(wrapper.find("#m-properties-panel").html()).toContain(
      notificationVehicle.operatorLastName
    )
  })

  test("if a vehicle from a notification is loading, show nothing", () => {
    const notification: Notification = { runIds: ["run_id"] } as Notification
    const state: State = { ...initialState, selectedNotification: notification }
    const wrapper = mount(
      <StateDispatchProvider state={state} dispatch={jest.fn()}>
        <VehicleForNotificationProvider vehicleForNotification={undefined}>
          <LadderPage />
        </VehicleForNotificationProvider>
      </StateDispatchProvider>
    )
    expect(wrapper.find("#m-properties-panel").exists()).toBeFalsy()
  })

  test("if a vehicle from a notification failed to load, show nothing", () => {
    const notification: Notification = { runIds: ["run_id"] } as Notification
    const state: State = { ...initialState, selectedNotification: notification }
    const wrapper = mount(
      <StateDispatchProvider state={state} dispatch={jest.fn()}>
        <VehicleForNotificationProvider vehicleForNotification={null}>
          <LadderPage />
        </VehicleForNotificationProvider>
      </StateDispatchProvider>
    )
    expect(wrapper.find("#m-properties-panel").exists()).toBeFalsy()
  })
})

describe("findRouteById", () => {
  test("finds a route in a list by its id", () => {
    expect(findRouteById(routes, "28")).toEqual({
      directionNames: { 0: "Outbound", 1: "Inbound" },
      id: "28",
      name: "28",
    })
  })

  test("returns undefined if the route isn't found", () => {
    expect(findRouteById(routes, "missing")).toEqual(undefined)
  })

  test("returns undefined if routes is null", () => {
    expect(findRouteById(null, "does not matter")).toEqual(undefined)
  })
})

describe("findSelectedVehicleOrGhost", () => {
  test("returns the requested vehicle if it is on the route", () => {
    expect(
      findSelectedVehicleOrGhost(vehiclesByRouteId, "on-route-39")
    ).toEqual({
      id: "on-route-39",
      routeStatus: "on_route",
    })
  })

  test("returns the requested vehicle if it is pulling out", () => {
    expect(
      findSelectedVehicleOrGhost(vehiclesByRouteId, "pulling-out-39")
    ).toEqual({
      id: "pulling-out-39",
      routeStatus: "pulling_out",
    })
  })

  test("returns the requested vehicle if it is a ghost bus", () => {
    expect(findSelectedVehicleOrGhost(vehiclesByRouteId, "ghost-39")).toEqual({
      id: "ghost-39",
    })
  })

  test("returns undefined if the vehicle is not found", () => {
    expect(
      findSelectedVehicleOrGhost(vehiclesByRouteId, "missing-23")
    ).toBeUndefined()
  })

  test("returns undefined if selectedVehicleId is undefined", () => {
    expect(
      findSelectedVehicleOrGhost(vehiclesByRouteId, undefined)
    ).toBeUndefined()
  })
})

const routes: Route[] = [
  { id: "1", directionNames: { 0: "Outbound", 1: "Inbound" }, name: "1" },
  { id: "28", directionNames: { 0: "Outbound", 1: "Inbound" }, name: "28" },
]
const timepointsByRouteId: TimepointsByRouteId = {
  "1": [
    { id: "WASMA", name: "WASMA Name" },
    { id: "MELWA", name: "MELWA Name" },
    { id: "HHGAT", name: "HHGAT Name" },
  ],
  "28": [
    { id: "MATPN", name: "MATPN Name" },
    { id: "WELLH", name: "WELLH Name" },
    { id: "MORTN", name: "MORTN Name" },
  ],
  "71": undefined,
  "73": null,
}

const vehiclesByRouteId: ByRouteId<VehicleOrGhost[]> = {
  "23": [
    {
      id: "on-route-23",
      routeStatus: "on_route",
    } as Vehicle,
    {
      id: "pulling-out-23",
      routeStatus: "pulling_out",
    } as Vehicle,
    {
      id: "ghost-23",
    } as Ghost,
  ],
  "39": [
    {
      id: "on-route-39",
      routeStatus: "on_route",
    } as Vehicle,
    {
      id: "pulling-out-39",
      routeStatus: "pulling_out",
    } as Vehicle,
    {
      id: "ghost-39",
    } as Ghost,
  ],
}

const notificationVehicle: Vehicle = {
  id: "id",
  label: "label",
  runId: "run",
  timestamp: 123,
  latitude: 0,
  longitude: 0,
  directionId: 0,
  routeId: "route",
  tripId: "t1",
  headsign: "Forest Hills",
  viaVariant: "X",
  operatorId: "op1",
  operatorFirstName: "PATTI",
  operatorLastName: "SMITH",
  operatorLogonTime: new Date("2018-08-15T13:38:21.000Z"),
  bearing: 33,
  blockId: "block-1",
  headwaySecs: 859.1,
  headwaySpacing: HeadwaySpacing.Ok,
  previousVehicleId: "v2",
  scheduleAdherenceSecs: 0,
  scheduledHeadwaySecs: 120,
  isShuttle: false,
  isOverload: false,
  isOffCourse: false,
  isRevenue: true,
  layoverDepartureTime: null,
  dataDiscrepancies: [
    {
      attribute: "trip_id",
      sources: [
        {
          id: "swiftly",
          value: "swiftly-trip-id",
        },
        {
          id: "busloc",
          value: "busloc-trip-id",
        },
      ],
    },
  ],
  stopStatus: {
    stopId: "s1",
    stopName: "Stop Name",
  },
  timepointStatus: {
    fractionUntilTimepoint: 0.5,
    timepointId: "tp1",
  },
  scheduledLocation: null,
  routeStatus: "on_route",
  endOfTripType: "another_trip",
  blockWaivers: [],
  crowding: null,
}
