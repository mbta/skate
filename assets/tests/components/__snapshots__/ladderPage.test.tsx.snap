// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LadderPage renders the empty state 1`] = `
<div
  className="m-ladder-page"
>
  <div
    className="m-picker-container visible"
  >
    <div
      className="m-picker-container__tab"
    >
      <button
        className="m-picker-container__tab-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-route-picker"
    >
      <ul
        className="m-route-picker__selected-routes"
      />
      <div
        className="m-route-filter"
      >
        <select
          className="m-route-filter__type"
          onChange={[Function]}
          value="id"
        >
          <option
            value="id"
          >
            Route ID
          </option>
        </select>
        <div
          className="m-route-filter__text"
        >
          <input
            className="m-route-filter__input"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value=""
          />
          <button
            className="m-route-filter__clear"
            onClick={[Function]}
          >
            <span
              className=""
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </button>
        </div>
      </div>
      loading...
    </div>
  </div>
  <div
    className="m-route-ladders"
  />
</div>
`;

exports[`LadderPage renders with routes 1`] = `
<div
  className="m-ladder-page"
>
  <div
    className="m-picker-container visible"
  >
    <div
      className="m-picker-container__tab"
    >
      <button
        className="m-picker-container__tab-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-route-picker"
    >
      <ul
        className="m-route-picker__selected-routes"
      >
        <li>
          <button
            className="m-route-picker__selected-routes-button"
            onClick={[Function]}
          >
            1
          </button>
        </li>
      </ul>
      <div
        className="m-route-filter"
      >
        <select
          className="m-route-filter__type"
          onChange={[Function]}
          value="id"
        >
          <option
            value="id"
          >
            Route ID
          </option>
        </select>
        <div
          className="m-route-filter__text"
        >
          <input
            className="m-route-filter__input"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value=""
          />
          <button
            className="m-route-filter__clear"
            onClick={[Function]}
          >
            <span
              className=""
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </button>
        </div>
      </div>
      <ul
        className="m-route-picker__route-list"
      >
        <li>
          <button
            className="m-route-picker__route-list-button m-route-picker__route-list-button--selected"
            onClick={[Function]}
          >
            1
          </button>
        </li>
        <li>
          <button
            className="m-route-picker__route-list-button "
            onClick={[Function]}
          >
            28
          </button>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="m-route-ladders"
  >
    <div
      className="m-route-ladder__header"
    >
      <button
        className="m-close-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
      <div
        className="m-route-ladder__route-name"
      >
        1
      </div>
    </div>
    <div
      className="m-route-ladder__controls"
    >
      <button
        className="m-route-ladder__reverse"
        onClick={[Function]}
      >
        <span
          className="m-route-ladder__reverse-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        Reverse
      </button>
    </div>
    loading...
  </div>
</div>
`;

exports[`LadderPage renders with selectedRoutes in different order than routes data 1`] = `
<div
  className="m-ladder-page"
>
  <div
    className="m-picker-container visible"
  >
    <div
      className="m-picker-container__tab"
    >
      <button
        className="m-picker-container__tab-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-route-picker"
    >
      <ul
        className="m-route-picker__selected-routes"
      >
        <li>
          <button
            className="m-route-picker__selected-routes-button"
            onClick={[Function]}
          >
            28
          </button>
        </li>
        <li>
          <button
            className="m-route-picker__selected-routes-button"
            onClick={[Function]}
          >
            1
          </button>
        </li>
      </ul>
      <div
        className="m-route-filter"
      >
        <select
          className="m-route-filter__type"
          onChange={[Function]}
          value="id"
        >
          <option
            value="id"
          >
            Route ID
          </option>
        </select>
        <div
          className="m-route-filter__text"
        >
          <input
            className="m-route-filter__input"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value=""
          />
          <button
            className="m-route-filter__clear"
            onClick={[Function]}
          >
            <span
              className=""
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </button>
        </div>
      </div>
      <ul
        className="m-route-picker__route-list"
      >
        <li>
          <button
            className="m-route-picker__route-list-button m-route-picker__route-list-button--selected"
            onClick={[Function]}
          >
            1
          </button>
        </li>
        <li>
          <button
            className="m-route-picker__route-list-button m-route-picker__route-list-button--selected"
            onClick={[Function]}
          >
            28
          </button>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="m-route-ladders"
  >
    <div
      className="m-route-ladder__header"
    >
      <button
        className="m-close-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
      <div
        className="m-route-ladder__route-name"
      >
        28
      </div>
    </div>
    <div
      className="m-route-ladder__controls"
    >
      <button
        className="m-route-ladder__reverse"
        onClick={[Function]}
      >
        <span
          className="m-route-ladder__reverse-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        Reverse
      </button>
    </div>
    loading...
    <div
      className="m-route-ladder__header"
    >
      <button
        className="m-close-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
      <div
        className="m-route-ladder__route-name"
      >
        1
      </div>
    </div>
    <div
      className="m-route-ladder__controls"
    >
      <button
        className="m-route-ladder__reverse"
        onClick={[Function]}
      >
        <span
          className="m-route-ladder__reverse-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        Reverse
      </button>
    </div>
    loading...
  </div>
</div>
`;

exports[`LadderPage renders with timepoints 1`] = `
<div
  className="m-ladder-page"
>
  <div
    className="m-picker-container visible"
  >
    <div
      className="m-picker-container__tab"
    >
      <button
        className="m-picker-container__tab-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-route-picker"
    >
      <ul
        className="m-route-picker__selected-routes"
      />
      <div
        className="m-route-filter"
      >
        <select
          className="m-route-filter__type"
          onChange={[Function]}
          value="id"
        >
          <option
            value="id"
          >
            Route ID
          </option>
        </select>
        <div
          className="m-route-filter__text"
        >
          <input
            className="m-route-filter__input"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value=""
          />
          <button
            className="m-route-filter__clear"
            onClick={[Function]}
          >
            <span
              className=""
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </button>
        </div>
      </div>
      <ul
        className="m-route-picker__route-list"
      >
        <li>
          <button
            className="m-route-picker__route-list-button "
            onClick={[Function]}
          >
            1
          </button>
        </li>
        <li>
          <button
            className="m-route-picker__route-list-button "
            onClick={[Function]}
          >
            28
          </button>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="m-route-ladders"
  />
</div>
`;
